File: leetcode_cpp/balanced_binary_tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:

    int getHeight(TreeNode *node) {
        if (!node) {
            return 0;
        }
        else if (!node -> left & !node -> right) {
            return 1;
        }
        int left = getHeight(node -> left), right = getHeight(node -> right);

        if (left == -1 | right == -1) {
            return -1;
        }
        else if (abs(left - right) <= 1) {
            return max(left, right) + 1;
        }
        else {
            return -1;
        }
    }

    bool isBalanced(TreeNode *root) {
        return !(getHeight(root) == -1);
    }
};

File: leetcode_python/balanced_binary_tree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a boolean

    def getHeight(self, node):
        if node is None:
            return 0
        elif node.left is None and node.right is None:
            return 1

        left = self.getHeight(node.left)
        right = self.getHeight(node.right)

        if left == -1 or right == -1:
            return -1
        elif abs(left - right) <= 1:
            return max(left, right) + 1
        else:
            return -1

    def isBalanced(self, root):
        return not(self.getHeight(root) == -1)
